opal/mca/threads/mutex.h

```diff
@@ -48,52 +51,127 @@ typedef struct opal_mutex_t opal_recursive_mutex_t;
 
 #include MCA_threads_mutex_base_include_HEADER
 
-OBJ_CLASS_DECLARATION(opal_mutex_t);
-OBJ_CLASS_DECLARATION(opal_recursive_mutex_t);
+struct opal_mutex_t {
+    opal_object_t super;
+    opal_thread_internal_mutex_t m_lock;
+#if OPAL_ENABLE_DEBUG
+    int m_lock_debug;
+    const char *m_lock_file;
+    int m_lock_line;
+#endif
+    opal_atomic_lock_t m_lock_atomic;
+};
+
+OPAL_DECLSPEC OBJ_CLASS_DECLARATION(opal_mutex_t);
+OPAL_DECLSPEC OBJ_CLASS_DECLARATION(opal_recursive_mutex_t);
+
+#if OPAL_ENABLE_DEBUG
+#    define OPAL_MUTEX_STATIC_INIT                                                         \
+        {                                                                                  \
+            .super = OPAL_OBJ_STATIC_INIT(opal_mutex_t),                                   \
+            .m_lock = OPAL_THREAD_INTERNAL_MUTEX_INITIALIZER, .m_lock_debug = 0,           \
+            .m_lock_file = NULL, .m_lock_line = 0, .m_lock_atomic = OPAL_ATOMIC_LOCK_INIT, \
+        }
+#else
+#    define OPAL_MUTEX_STATIC_INIT                            \
+        {                                                     \
+            .super = OPAL_OBJ_STATIC_INIT(opal_mutex_t),      \
+            .m_lock = OPAL_THREAD_INTERNAL_MUTEX_INITIALIZER, \
+            .m_lock_atomic = OPAL_ATOMIC_LOCK_INIT,           \
+        }
+#endif
+
+#if defined(OPAL_THREAD_INTERNAL_RECURSIVE_MUTEX_INITIALIZER)
+#    if OPAL_ENABLE_DEBUG
+#        define OPAL_RECURSIVE_MUTEX_STATIC_INIT                                               \
+            {                                                                                  \
+                .super = OPAL_OBJ_STATIC_INIT(opal_mutex_t),                                   \
+                .m_lock = OPAL_THREAD_INTERNAL_RECURSIVE_MUTEX_INITIALIZER, .m_lock_debug = 0, \
+                .m_lock_file = NULL, .m_lock_line = 0, .m_lock_atomic = OPAL_ATOMIC_LOCK_INIT, \
+            }
+#    else
+#        define OPAL_RECURSIVE_MUTEX_STATIC_INIT                            \
+            {                                                               \
+                .super = OPAL_OBJ_STATIC_INIT(opal_mutex_t),                \
+                .m_lock = OPAL_THREAD_INTERNAL_RECURSIVE_MUTEX_INITIALIZER, \
+                .m_lock_atomic = OPAL_ATOMIC_LOCK_INIT,                     \
+            }
+#    endif
+#endif /* OPAL_THREAD_INTERNAL_RECURSIVE_MUTEX_INITIALIZER */
 
 /**
  * Try to acquire a mutex.
  *
  * @param mutex         Address of the mutex.
  * @return              0 if the mutex was acquired, 1 otherwise.
  */
-static inline int opal_mutex_trylock(opal_mutex_t *mutex);
+static inline int opal_mutex_trylock(opal_mutex_t *mutex)
```

Comment:
Are you sure these are properly inlined such we don't get one function call ?

Reply:
I cannot 100% guarantee such for all the existing C compilers, but in my experience, GCC/XLC/ICC/Clang will inline-expand small leaf functions unless you set `-O0`.